### DISCUSSION TITLE
How to customize standard json schema generation based on jaxb classes to set base64 contentEncoding

### DISCUSSION THREAD

**Discussion by theseeker58** [victools/jsonschema-generator#451] [Q&A] [answered]
Hello there, I am trying to generate a json schema starting from jaxb classes generated through xjc and xjb custom bindings

This is the pom fragment
```xml
        <configuration>
          <classNames>org.foo.bar.Document</classNames>
          <schemaVersion>DRAFT_2020_12</schemaVersion>
          <schemaFileName>document-schema.json</schemaFileName>
          <schemaFilePath>target/generated-sources/json-schema</schemaFilePath>
          <modules>
            <module>
              <name>JakartaValidation</name>
              <options>
                <option>NOT_NULLABLE_FIELD_IS_REQUIRED</option>
                <option>INCLUDE_PATTERN_EXPRESSIONS</option>
              </options>
            </module>
          </modules>
        </configuration>

```
Everything works perfectly but a XML tag defined xs:base64Binary in the xsd schema.
xjc maps it to an array of primitive bytes like this

```java
    @XmlElement(name = "Nclsr", required = true)
    @JsonProperty(required = true)
    @NotNull
    @Size(min = 1, max = 10485760)
    protected byte[] enclosure;
```
 and the json schema generator accordingly generates this (an array of string)

```json
                         "enclosure" : {
                            "minItems" : 1,
                            "maxItems" : 10485760,
                            "type" : "array",
                            "items" : {
                              "type" : "string"
                            }
                          },
```
Is there a way to customize the schema generation to get the following expected definition with content encoding and min/maxLength of a single string instead of an array of strings with min/maxItems? 

```json
                         "enclosure" : {
                            "minLength" : 1,
                            "maxLength" : 10485760,
                            "type" : "string",
                            "items" : {
                              "type" : "string",
                              "contentEncoding" : "base64"
                            }
                          },
```
possibly inside maven pom.

I have read the discussion[ How to programmatically generate JsonSchema representing binary content in Java #271](https://github.com/victools/jsonschema-generator/discussions/271),
but unfortunately I need not to tamper the original xsd (it's an ISO XML 20022 standard) and the enclosure tag is inside a complexType and therefore xjc generate the enclosure property as an array of bytes inside another pojo without a wrapper class as suggested in that discussion.

For the record that property will accept a URL base64 encoded and accordingly on the interbank network space the maxLength is quite lesser. But it's not a problem as the restriction won't be enforced by the json schema.

Many thanks

**Comment by CarstenWickner** [comment, marked as answer]
Hi @theseeker58,

This is certainly possible. I've created an example under #452 for the required custom configuration:
```java
configBuilder.forFields()
        .withTargetTypeOverridesResolver(this::treatByteArrayAsString)
        .withInstanceAttributeOverride(this::addContentEncodingForByteArray);
```
```java
private List<ResolvedType> treatByteArrayAsString(MemberScope<?, ?> scope) {
    ResolvedType type = scope.getType();
    if (type != null && type.getErasedType().equals(byte[].class)) {
        return Collections.singletonList(scope.getContext().resolve(String.class));
    }
    return null;
}

private void addContentEncodingForByteArray(ObjectNode schema, MemberScope<?, ?> scope, SchemaGenerationContext context) {
    ResolvedType declaredType = scope.getDeclaredType();
    if (declaredType != null && declaredType.getErasedType().equals(byte[].class)) {
        schema.put("contentEncoding", "base64");
    }
}
```

For including such custom configuration in your Maven plugin configuration, you can follow the corresponding entry in the documentation: https://victools.github.io/jsonschema-generator/#further-configurations-through-modules
In short: wrap the configuration into an implementation of the `Module` interface and then add that to your Maven plugin configuration's `<modules>` list.

**Reply by theseeker58** [reply]
Thank you @CarstenWickner,
for simplicity I have tried first as you answered. Then I have put the code in a ByteArrayConverter class that implements Module interface
```
public class ByteArrayConverter implements Module {

  @Override
  public void applyToConfigBuilder(SchemaGeneratorConfigBuilder builder) {
    builder.forFields()
        .withTargetTypeOverridesResolver(this::treatByteArrayAsString)
        .withInstanceAttributeOverride(this::addContentEncodingForByteArray);
  }

  private List<ResolvedType> treatByteArrayAsString(MemberScope<?, ?> scope) {
    ResolvedType type = scope.getType();
    if (type != null && type.getErasedType().equals(byte[].class)) {
      return Collections.singletonList(scope.getContext().resolve(String.class));
    }
    return null;
  }

  private void addContentEncodingForByteArray(ObjectNode schema,
      MemberScope<?, ?> scope, SchemaGenerationContext context) {
    ResolvedType declaredType = scope.getDeclaredType();
    if (declaredType != null && declaredType.getErasedType().equals(byte[].class)) {
      schema
          .put("contentEncoding", "base64")
          .put("contentMediaType", "text/plain");

    }
  }
}
```
and finally modified POM file under `<modules>` tag as follow:
```
              <module>
                <className>ByteArrayConverter</className>
              </module>
```
Added contentMediaType as well.

Thanks again for your support
